third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,885 @@
+        }
+        // Rename local variables in the pattern.
+        let opt_new_pat = pat.replace_vars(self.remapped_symbol.borrow_map());
+        if let Some(new_pat) = &opt_new_pat {
+            if SIMPLIFIER_DEBUG {
+                eprintln!(
+                    "Entering block {}, Pat was {}, renamed to {}, remapped_symbol map is {:#?}",
+                    id.as_usize(),
+                    pat.to_string(self.env, &TypeDisplayContext::new(self.env)),
+                    new_pat.to_string(self.env, &TypeDisplayContext::new(self.env)),
+                    self.remapped_symbol.borrow_map(),
+                );
+            }
+        } else {
+            if SIMPLIFIER_DEBUG {
+                eprintln!(
+                    "Entering block {}, Pat was {}, renamed to None, remapped_symbol map is {:#?}",
+                    id.as_usize(),
+                    pat.to_string(self.env, &TypeDisplayContext::new(self.env)),
+                    self.remapped_symbol.borrow_map(),
+                );
+            }
+        }
+        opt_new_pat
+    }
+
+    // Note that `rewrite_block` is called *after* `rewrite_exit_scope`.
+    fn rewrite_block(
+        &mut self,
+        id: NodeId,
+        pat: &Pattern,
+        opt_binding: &Option<Exp>,
+        body: &Exp,
+    ) -> Option<Exp> {
+        if let Some(exp) = opt_binding {
+            let pat_id = pat.node_id();
+            let exp_id = exp.node_id();
+            let pat_type = self.env.get_node_type(pat_id);
+            let exp_type = self.env.get_node_type(exp_id);
+            let type_display_context = TypeDisplayContext::new(self.env);
+            if SIMPLIFIER_DEBUG {
+                eprintln!(
+                    "Starting rewrite_block(id={}, pat={}, opt_binding={}, body={}, pat_type={}, exp_type={}, {})",
+                    id.as_usize(),
+                    pat.to_string(self.env, &type_display_context),
+                    exp.display_verbose(self.env),
+                    body.display_verbose(self.env),
+                    pat_type.display(&type_display_context),
+                    exp_type.display(&type_display_context),
+                    if pat_type == exp_type { "MATCHES" } else { "NO MATCH" },
+                );
+            }
+        } else {
+            if SIMPLIFIER_DEBUG {
+                eprintln!(
+                    "Starting rewrite_block(id={}, pat={}, opt_binding={}, body={})",
+                    id.as_usize(),
+                    pat.to_string(self.env, &TypeDisplayContext::new(self.env)),
+                    "None",
+                    body.display_verbose(self.env)
+                );
+            }
+        }
+
+        // Simplify binding:
+        //   A few ideas for simplification which are implemented below:
+        //     (1) if no binding, then simplify to just the body.
+        //     (2) if all pattern vars are unused in body and binding is side-effect free, again return body.
```

Comment:
This places the requirement that unused variables check is performed before the AST simplification: we might want to document that explicitly in the module documentation.